lib/logflare/source/bigquery/buffer_producer.ex

```diff
@@ -33,62 +31,34 @@ defmodule Logflare.Source.BigQuery.BufferProducer do
     {:noreply, [], state}
   end
 
-  def handle_receive_messages(%{receive_timer: nil, demand: demand} = state) when demand > 0 do
-    messages = receive_messages_from_buffer(state, demand)
-    new_demand = demand - length(messages)
-
-    receive_timer =
-      case {messages, new_demand} do
-        {[], _} -> schedule_receive_messages(state.receive_interval)
-        {_, 0} -> nil
-        _ -> schedule_receive_messages(0)
-      end
-
-    {:noreply, messages, %{state | demand: new_demand, receive_timer: receive_timer}}
-  end
-
-  def handle_receive_messages(state) do
-    {:noreply, [], state}
-  end
-
-  def ack(table, successful, unsuccessful) do
+  @spec ack(atom(), [%Broadway.Message{}], [%Broadway.Message{}]) :: :ok
+  def ack(source_id, successful, unsuccessful) when is_atom(source_id) do
     Enum.each(successful, fn %{data: %LE{}} = message ->
-      Buffer.ack(table, message.data.id)
+      Buffer.ack(source_id, message.data.id)
     end)
 
     Enum.each(unsuccessful, fn %{data: %LE{}} = message ->
-      {:ok, le} = Buffer.ack(table, message.data.id)
-      Buffer.push(le)
+      Buffer.ack(source_id, message.data.id)
     end)
-  end
 
-  @impl true
-  def terminate(reason, _state) do
-    Logger.info("Going Down - #{inspect(reason)} - #{__MODULE__}")
-    reason
+    :ok
   end
 
-  defp receive_messages_from_buffer(%{source_id: source_id}, total_demand) do
-    events =
-      source_id
-      |> Buffer.pop_many(total_demand)
-
-    for e <- events do
-      %Broadway.Message{
-        data: e,
-        acknowledger: {__MODULE__, source_id, "no idea what this does"}
-      }
-    end
+  defp handle_receive_messages(
+         %{source_id: source_id, receive_timer: nil, demand: demand} = state
+       )
+       when demand > 0 do
+    Buffer.pop(source_id, demand)
```

Comment:
Is this function in use looks private and doesn't seem to be called.?

I don't quite get why we do this for new demand, `pop/2` will decrease the counter but won't actually send any messages to Broadway's buffer, so in effect this would result in a double decrement, since ack will decrease the counter by 1 and then the receive message will decrease it again by demanded amount.

Would be good to test for the Buffer (BufferCounter) behaviour as the log event goes through the broadway pipeline as well as after it gets acked + new demand.

Reply:
The BufferCounter keeps the popped, pushed and ackd in state only incrementing them. 

`get_count` just does the math when to subtract the ackd from pushed.  If everything gets into BigQuery pushed minus ackd should be 0. 

I was the `pop` should actually never get called because Broadway should never send the producer demand greater than 0.

Reply:
Well I just pulled it out.